packages/substrate-codecs/test/AccountId.spec.ts

```diff
@@ -0,0 +1,27 @@
+import { expect, test } from "vitest"
+import { bytesToHex } from "@noble/hashes/utils"
+
+import { AccountId } from "../src/AccountId"
+
+test.each([
+  [
+    0,
+    "Polkadot",
+    "15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5",
+    "d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d",
+  ],
+  [
+    42,
+    "Substrate",
+    "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
+    "d43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d",
+  ],
```

Comment:
@josepot  There are some other test cases for ss58 addresses with different payload and checksum sizes https://github.com/paritytech/capi-old/blob/15fbc54133c1e1e986156400708ea5a4f93672e2/crypto/ss58.test.ts#L23-L106

What would you suggest to do?
Shall we only support 32/33 byte payload sizes with 2 byte checksums?

Reply:
Hey @kratico, thanks for working on this with me! We've stumbled into a bit of a chicken-and-egg situation with the `AccountId` function. I know I asked you to write tests, but I didn't properly explain how it should work. My bad! 😅 🙏

Let's clarify that now. Here's the main idea of the `AccountId` function:

```ts
type AccountIdFn = (ss58Format: number) => Codec<string>
```

`AccountId` is a Codec creator. You give it a number representing the `ss58Format`, and it hands back a Codec to encode/decode Substrate public keys using that format. Here's what we expect from it:

- **Decoding**: It takes **only** the bytes of the public key and turns them into [a Substrate ss58 formatted address](https://docs.substrate.io/reference/address-formats/). I believe this part is working well, but if you find a test that proves otherwise, that's excellent—we can fix it.

- **Encoding**: Here's where it gets interesting:
  * It accepts a string with a Substrate ss58 formatted address and returns only the bytes of the public key.
  * But it also needs to verify that:
    * The input is a valid base58 string.
    * It's a valid ss58 formatted address, checking that:
      * The `ss58Format` extracted from the input matches the one given when creating the Codec.
      * The checksum is good to go.

If some of those verifications fail, then the encoder should throw an error. We can talk later about whether we need special error types for these situations, but for now, we just need to make sure it throws when something's off. This is where I know we need to improve. Right now, the encoder only checks if it's a valid base58 string, so it would be awesome to add tests (and possibly fixes?) that show how the encoder behaves when it receives an address formatted with a different `ss58Format`, or if the checksum is invalid.

So that's what we're aiming for. I appreciate your help in getting this sorted out, and I'm here if you have any questions or need further details. Let's make this thing rock! 🎸

Reply:
I think we should be good supporting 32-byte payloads for `AccountId` (or `AccountId32`) because I haven't seen a different byte length in Substrate.

These are the ss58 address format docs https://docs.substrate.io/reference/address-formats/#address-length-in-bytes

I also found some [substrate tests](https://github.com/paritytech/substrate/blob/51b2f0ed6af8dd4facb18f1a489e192fd0673f7b/primitives/core/src/ecdsa.rs#L680-L712) for a 33-byte [ECDSA](https://en.bitcoin.it/wiki/Elliptic_Curve_Digital_Signature_Algorithm) compressed public key. The first byte is a prefix so, I guess that the remaining 32 would be the public key that can be used as `AccountId`.  
Nevertheless, this is using ss58 to encode a 33-byte public key.  
Later, we should write a test to send an extrinsic signed with ECDSA and check the above. 

Leaving some links in case we need to revisit this in the future
- other implementations supporting different payloads and checksums
  - [py-scale-codec](https://github.com/polkascan/py-scale-codec/blob/045b34813727f25602c2e5d2440fffc621d55680/scalecodec/utils/ss58.py#L121-L128)
  - [subscan-essentials](https://github.com/subscan-explorer/subscan-essentials/blob/master/util/ss58/ss58.go)
- Previous ss58 implementation
  - https://github.com/paritytech/capi-old/issues/619
  - https://github.com/paritytech/capi-old/pull/751